src/mlpack/tests/main_tests/hmm_train_test.cpp

```diff
@@ -0,0 +1,404 @@
+/**
+ * @file hmm_training_tests.cpp
+ * @author Daivik Nema
+ *
+ * Test mlpackMain() of hmm_train_main.cpp.
+ */
+#include <string>
+#include <fstream>
+
+#define BINDING_TYPE BINDING_TYPE_TEST
+static const std::string testName = "HMMTrain";
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include "test_helper.hpp"
+#include <mlpack/methods/hmm/hmm_train_main.cpp>
+#include <mlpack/methods/hmm/hmm_model.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
+
+using namespace mlpack;
+
+struct HMMTrainMainTestFixture
+{
+ public:
+  HMMTrainMainTestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~HMMTrainMainTestFixture()
+  {
+    // Clear the settings.
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(HMMTrainMainTest, HMMTrainMainTestFixture);
+
+inline void FileExists(std::string fileName)
+{
+  ifstream ifp(fileName);
+  if (!ifp.good())
+    BOOST_FAIL("Bad stream " + fileName);
+  ifp.close();
+}
+
+inline bool ApproximatelyEqual(HMMModel& h1,
+                               HMMModel& h2,
+                               double tolerance)
+{
+  if (h1.Type() != h2.Type())
+    return false;
+  HMMType  hmmType = h1.Type();
+  bool transitionEqual = false;
+  bool emissionEqual = false;
+  bool initialEqual = false;
+  if (hmmType ==  DiscreteHMM)
+  {
+    transitionEqual = approx_equal(
```

Comment:
Hmm, this is unfortunate... the minimum version of Armadillo for mlpack is 6.500, but `approx_equal()` isn't available until 6.700.  I don't think it's realistic to bump the minimum version of Armadillo for this, so there are two things we could do:

 * Backport `fn_approx_equal.hpp`, by taking `fn_approx_equal.hpp` from the newest version of Armadillo, putting some `#if` guards around it so that it only gets included for sufficiently old versions of Armadillo (older than 6.700), and adding it to `arma_extend.hpp`.  Here's a commit where I did exactly that for a sparse matrix subtraction operator: https://github.com/mlpack/mlpack/commit/62def05b672f5a44dbdc2c721b522ac4bf10c7d0#diff-b3e69ecff96ebaae403c243f71c038db

 * Avoid `approx_equal` and write a loop by hand or use `CheckMatrices()` from `test_tools.hpp`.

I think `CheckMatrices` may be the easiest way to go, but it's up to you how you want to do it.